Issue #3654: add documentation for kerberos authentication (#4134)
apache/pulsar: site2/docs/security-kerberos.md

---
id: security-kerberos
title: Authentication using Kerberos
sidebar_label: Authentication using Kerberos
---

[Kerberos](https://web.mit.edu/kerberos/) is a network authentication protocol. It is designed to provide strong authentication for client/server applications by using secret-key cryptography. 

In Pulsar, we use Kerberos with [SASL](https://en.wikipedia.org/wiki/Simple_Authentication_and_Security_Layer) as a choice for authentication. While in SASL, Pulsar uses the [Java Authentication and Authorization Service (JAAS)](https://en.wikipedia.org/wiki/Java_Authentication_and_Authorization_Service) for SASL configuration. You must provide JAAS configurations for Kerberos authentication. 

In this documentation, We will start with a general description of how to configure `Kerberos` with `SASL` between Pulsar clients and brokers, followed by some operational details. Then comparing with it, we will show how to configure for Pulsar Proxy.

## Configuration for Kerberos between Client and Broker

### Prerequisites

To begin, you need to set up(or already have) a [Key Distribution Center](KDC) configured and running. 

> If your organization is already using a Kerberos server (for example, by using `Active Directory`), there is no need to install a new server just for Pulsar. Otherwise you will need to install one, your Linux vendor likely has packages for `Kerberos` and a short guide on how to install and configure it ([Ubuntu](https://help.ubuntu.com/community/Kerberos),

Reviewer: ```suggestion
If your organization is using a Kerberos server (for example, `Active Directory`), there is no need to install a new server for Pulsar. Otherwise, you need to install one. Your Linux vendor has packages for `Kerberos` and a short guide on how to install and configure it ([Ubuntu](https://help.ubuntu.com/community/Kerberos),
```

[Redhat](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Managing_Smart_Cards/installing-kerberos.html)).

> Note that if you are using Oracle Java, you will need to download JCE policy files for your Java version and copy them to `$JAVA_HOME/jre/lib/security`.

Reviewer: ```suggestion
Note that if you are using Oracle Java, you need to download JCE policy files and copy them to the `$JAVA_HOME/jre/lib/security` directory.
```

#### Kerberos Principals

> If you are using the organization’s Kerberos or Active Directory server, ask your Kerberos administrator for a principal for each Brokers in your cluster and for every operating system user that will access Pulsar with Kerberos authentication(via clients and tools).

Reviewer: The sentence is a little long, maybe you can use some short sentences, it's easier to understand.

> If you have installed your own Kerberos, you will need to create these principals yourself using the following commands:

Reviewer: ```suggestion
If you have installed your own Kerberos, you can create these principals with the following commands:
```

```shell
### add Principals for broker
sudo /usr/sbin/kadmin.local -q 'addprinc -randkey broker/{hostname}@{REALM}'
sudo /usr/sbin/kadmin.local -q "ktadd -k /etc/security/keytabs/{broker-keytabname}.keytab broker/{hostname}@{REALM}"
### add Principals for client
sudo /usr/sbin/kadmin.local -q 'addprinc -randkey client/{hostname}@{REALM}'
sudo /usr/sbin/kadmin.local -q "ktadd -k /etc/security/keytabs/{client-keytabname}.keytab client/{hostname}@{REALM}"
```
Note that it is a *Kerberos* requirement that all your hosts can be resolved with their FQDNs.

#### Configure how to connect to KDC

> User need to specify the path to the `krb5.conf` file for both client and broker side. `krb5.conf` indicate the default Realm and KDC information. see [JDK’s Kerberos Requirements](https://docs.oracle.com/javase/8/docs/technotes/guides/security/jgss/tutorials/KerberosReq.html) for more details):

Reviewer: ```suggestion
You need to specify the path to the `krb5.conf` file for both client and broker side. The `krb5.conf` file indicates the default Realm and KDC information. See [JDK’s Kerberos Requirements](https://docs.oracle.com/javase/8/docs/technotes/guides/security/jgss/tutorials/KerberosReq.html) for more details.
```

```shell
-Djava.security.krb5.conf=/etc/pulsar/krb5.conf
```

> Here is an example of krb5.conf file:

Reviewer: ```suggestion
Here is an example of the krb5.conf file:
```

> Here is an example of krb5.conf file:

Reviewer 2: Usually machines configure  with kerberos already have a system wide configuration and this configuration is optional.

> In this configuration file, it provides `EXAMPLE.COM` as default realm; set `udp_preference_limit = 1` to force use TCP; set kdc server url `kdc = localhost:62037` for realm `EXAMPLE.COM `:

Reviewer: ```suggestion
In the configuration file, `EXAMPLE.COM` is the default realm; setting `udp_preference_limit = 1` is to use TCP by force; set kdc server url `kdc = localhost:62037` for realm `EXAMPLE.COM `:
```

```
[libdefaults]
 default_realm = EXAMPLE.COM

>  udp_preference_limit = 1

Reviewer 2: This is not needed.
It is only for our tests, to force tcp

[realms]
 EXAMPLE.COM  = {
  kdc = localhost:62037
 }
```

#### JAAS configuration file

> JAAS configuration file is needed for both client and broker side. It provides the section information that will use to connect KDC. Here is an example named `pulsar_jaas.conf`:

Reviewer: "It provides the section information that will use to connect KDC." 
Could we refine it as "It provides information used to connect KDC."? I'm not sure whether I've understood it correctly.

```
 PulsarBroker {
   com.sun.security.auth.module.Krb5LoginModule required
   useKeyTab=true
   storeKey=true
   useTicketCache=false
   keyTab="/etc/security/keytabs/pulsarbroker.keytab"
   principal="broker/localhost@EXAMPLE.COM";
};

 PulsarClient {
   com.sun.security.auth.module.Krb5LoginModule required
   useKeyTab=true
   storeKey=true
   useTicketCache=false
   keyTab="/etc/security/keytabs/pulsarclient.keytab"
   principal="client/localhost@EXAMPLE.COM";
};
```

> User need set the `JAAS` configuration file location as JVM parameter to client and broker. For example:

Reviewer: Could we refine it as follows?
You need to set the `JAAS` configuration file path as JVM parameter for client and broker.

```shell
    -Djava.security.auth.login.config=/etc/pulsar/pulsar_jaas.conf 
```

> In above `pulsar_jaas.conf` file 

Reviewer: ```suggestion
In the `pulsar_jaas.conf` file above
```
